pkg/middlewares/metrics/metrics.go

```diff
@@ -78,23 +77,22 @@ func WrapServiceHandler(ctx context.Context, registry metrics.Registry, serviceN
 }
 
 func (m *metricsMiddleware) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
-	labels := []string{"method", getMethod(req), "protocol", getRequestProtocol(req)}
+	// Initialize labels slice with correct size
+	labels := make([]string, 0, 6+len(m.baseLabels))
+	labels = append(labels, []string{"method", getMethod(req), "protocol", getRequestProtocol(req)}...)
 	labels = append(labels, m.baseLabels...)
 
-	openConns := atomic.AddInt64(&m.openConns, 1)
-	m.openConnsGauge.With(labels...).Set(float64(openConns))
-	defer func(labelValues []string) {
-		openConns := atomic.AddInt64(&m.openConns, -1)
-		m.openConnsGauge.With(labelValues...).Set(float64(openConns))
-	}(labels)
+	m.openConnsGauge.With(labels...).Add(1)
+	defer m.openConnsGauge.With(labels...).Add(-1)
```

Comment:
I restored the `defer` call instead of moving `m.openConnsGauge.With(labels...).Add(-1)` after `m.next.ServeHTTP(recorder, req)` because I've a scenario where, when doing a load testing of an SSE endpoint with rate limiting activated, we never go past `m.next.ServeHTTP(recorder, req)` and the Gauge is never decremented.

@mpl if you have an idea why I would be really interested.

Reply:
well, that suggests m.next.ServeHTTP never returns, either because of a deadlock or a blocked channel, does it not?